clientUtils/SqlFilterSExpression.ts

```diff
@@ -0,0 +1,515 @@
+import parse = require("s-expression")
+import pointer from "json8-pointer"
+import { isArray, tail, without } from "lodash"
+export type SExprAtom = string | String | SExprAtom[] // eslint-disable-line @typescript-eslint/ban-types
+
+export enum Arity {
+    nullary = "nullary",
+    unary = "unary",
+    binary = "binary",
+    nary = "nary",
+}
+
+const arityToNumberMap = new Map([
+    [Arity.nullary, 0],
+    [Arity.unary, 1],
+    [Arity.binary, 2],
+    [Arity.nary, undefined],
+])
+
+export interface Operation {
+    toSql(): string
+    toSExpr(): string
+    expressionType: ExpressionType
+}
+
+export abstract class NumericOperation implements Operation {
+    abstract toSql(): string
+    abstract toSExpr(): string
+    expressionType: ExpressionType = ExpressionType.numeric
+}
+
+export abstract class BooleanOperation implements Operation {
+    abstract toSql(): string
+    abstract toSExpr(): string
+    expressionType: ExpressionType = ExpressionType.boolean
+}
+
+export abstract class StringOperation implements Operation {
+    abstract toSql(): string
+    abstract toSExpr(): string
+    expressionType: ExpressionType = ExpressionType.string
+}
+
+export enum ExpressionType {
+    numeric = "numeric",
+    boolean = "boolean",
+    string = "string",
+    any = "any", // This is used for values we can't type check e.g. jsonpath queries that might return anything
+}
+
+export class BooleanAtom extends BooleanOperation {
+    constructor(v: boolean) {
+        super()
+        this.value = v
+    }
+    value: boolean
+    toSql(): string {
+        return this.value.toString()
+    }
+
+    toSExpr(): string {
+        return this.value.toString()
+    }
+}
+
+export class NumberAtom extends NumericOperation {
+    constructor(v: number) {
+        super()
+        this.value = v
+    }
+    value: number
+    toSql(): string {
+        return this.value.toString()
+    }
+
+    toSExpr(): string {
+        return this.value.toString()
+    }
+}
+
+export class StringAtom extends StringOperation {
+    constructor(v: string) {
+        super()
+        this.value = v
+    }
+    expressionType = ExpressionType.string
+    value: string
+    escapedValue(): string {
+        return this.value.toString().replace("'", "''")
+    }
+    toSql(): string {
+        return `'${this.escapedValue()}'` // escape single quotes to avoid SQL injection attacks :)
```

Comment:
Looking at [PHP's `mysqli_real_escape_string`](https://www.php.net/manual/en/mysqli.real-escape-string.php), it says
> Characters encoded are `NUL (ASCII 0), \n, \r, \, ', ", and Control-Z`.

So we should also ensure that there really cannot be any SQL Injection happening.

---
Your current implementation is definitely susceptible to `abc\'; DROP TABLE students; --`

Reply:
Very nice catch! I think if we add a replace from \ to \\ that should cover all cases though?